lte/gateway/c/li_agent/test/test_pdu_generator.cpp

```diff
@@ -0,0 +1,71 @@
+/**
+ * Copyright 2020 The Magma Authors.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <chrono>
+#include <thread>
+
+#include <netinet/ip.h>
+#include <net/ethernet.h>
+
+#include "Consts.h"
+#include "PDUGenerator.h"
+#include "LIAgentDMocks.h"
+
+#include <gtest/gtest.h>
+
+using ::testing::Test;
+
+namespace magma {
+
+class PDUGeneratorTest : public ::testing::Test {
+ protected:
+  virtual void SetUp() {
+    auto proxy_connector_p   = std::make_unique<MockProxyConnector>();
+    proxy_connector          = proxy_connector_p.get();
+    auto directoryd_client_p = std::make_unique<MockDirectorydClient>();
+    directoryd_client        = directoryd_client_p.get();
+
+    pkt_generator = std::make_unique<PDUGenerator>(
+        std::move(proxy_connector_p), std::move(directoryd_client_p),
+        PKT_DST_MAC, PKT_SRC_MAC);
+  }
+
+  MockProxyConnector* proxy_connector;
+  MockDirectorydClient* directoryd_client;
+  std::unique_ptr<PDUGenerator> pkt_generator;
+};
+
+TEST_F(PDUGeneratorTest, test_pdu_generator) {
+  struct pcap_pkthdr* phdr =
+      (struct pcap_pkthdr*) malloc(sizeof(struct pcap_pkthdr));
+  phdr->len       = sizeof(struct ether_header) + sizeof(struct ip);
+  phdr->ts.tv_sec = 92;
+  u_char* pdata =
+      (u_char*) malloc(sizeof(struct ether_header) + sizeof(struct ip));
+  struct ether_header* ethernetHeader = (struct ether_header*) pdata;
+  ethernetHeader->ether_type          = htons(ETHERTYPE_IP);
+  struct ip* ipHeader     = (struct ip*) (pdata + sizeof(struct ether_header));
+  ipHeader->ip_src.s_addr = 123;
+  ipHeader->ip_dst.s_addr = 1222787743;
+  pkt_generator->send_packet(phdr, pdata);
+  EXPECT_CALL(
```

Comment:
for some reason directoryd_client is not catching the call, if anyone knows why, lmk.
```
1/1 Test #1: test_pdu_generator ...............***Failed    0.01 sec
Running main() from gmock_main.cc
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from PDUGeneratorTest
[ RUN      ] PDUGeneratorTest.test_pdu_generator
02 00 02 00 28 00 00 00 22 00 00 00 05 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 5C 00 00 00 00 00 00 00 F0 87 9A 4E D7 55 00 00 3A 31 33 3A 08 00 3A 31 35 00 79 44 69 72 65 63 74 6F 7B 00 00 00 9F 42 E2 48 

GMOCK WARNING:
Uninteresting mock function call - returning directly.
    Function call: get_directoryd_xid_field(@0x7ffd6400e2a8 "123.0.0.0", 32-byte object <90-0D 9B-4E D7-55 00-00 50-87 9A-4E D7-55 00-00 64-5A 9D-4D D7-55 00-00 0B-5A 9D-4D D7-55 00-00>)
NOTE: You can safely ignore the above warning unless this call should not happen.  Do not suppress it by blindly adding an EXPECT_CALL() if you don't mean to enforce the call.  See https://github.com/google/googletest/blob/master/googlemock/docs/CookBook.md#knowing-when-to-expect for details.

GMOCK WARNING:
Uninteresting mock function call - returning directly.
    Function call: get_directoryd_xid_field(@0x7ffd6400e2c8 "159.66.226.72", 32-byte object <00-20 9B-4E D7-55 00-00 50-87 9A-4E D7-55 00-00 64-5A 9D-4D D7-55 00-00 0B-5A 9D-4D D7-55 00-00>)
NOTE: You can safely ignore the above warning unless this call should not happen.  Do not suppress it by blindly adding an EXPECT_CALL() if you don't mean to enforce the call.  See https://github.com/google/googletest/blob/master/googlemock/docs/CookBook.md#knowing-when-to-expect for details.
/home/vagrant/magma/lte/gateway/c/li_agent/test/test_pdu_generator.cpp:61: Failure
Actual function call count doesn't match EXPECT_CALL(*directoryd_client, get_directoryd_xid_field(testing::_, testing::_))...
         Expected: to be called once
           Actual: never called - unsatisfied and active
/home/vagrant/magma/lte/gateway/c/li_agent/test/test_pdu_generator.cpp:63: Failure
Actual function call count doesn't match EXPECT_CALL(*directoryd_client, get_directoryd_xid_field(testing::_, testing::_))...
         Expected: to be called once
           Actual: never called - unsatisfied and active
[  FAILED  ] PDUGeneratorTest.test_pdu_generator (0 ms)
[----------] 1 test from PDUGeneratorTest (0 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (0 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] PDUGeneratorTest.test_pdu_generator

 1 FAILED TEST

```

Reply:
Have you tried printing the DirectoryDMock pointers to make sure what you're asserting on and what the code calls are the same things?
`std::cout` is your friend for test logging.

Reply:
Hmm actually I checked locally and that doesn't seem to be the problem...